### PR TITLE
Fix: Improve hint when binding or rebinding variables omitting let.

### LINKED ISSUE #1580: Improve hint when binding or rebinding variables omitting `let`
Attempts of binding or rebinding variables without the `let` keyword should print a more obvious error message:

```gleam
pub fn main() {
 foo = 42
}
```

```
error: Syntax error
  ┌─ ./src/main.gleam:2:6
  │
2 │  foo = 42
  │      ^ I was not expecting this.

Expected one of: "}"
```

See: <https://johndoneth.github.io/gleam-playground/?s=A4VwRgBAZgdhC2BDAljAFASggbwFDQHsCIBeCAFgCZcBfIA%3D>

Ref: https://discord.com/channels/768594524158427167/814516237173784607/961748042481041409

### PR DESCRIPTION
**Pull request by Kn99HN**
- Closes #1580

### CODE REVIEW

**Comment by inoas** [comment]
I also noticed that this error message isn't optimal:

```gleam
pub fn main() {
  let foo
}
```

```
error: Syntax error
  ┌─ ./src/main.gleam:3:1
  │
3 │ }
  │ ^ I was not expecting this.

Expected one of: "="
```

https://johndoneth.github.io/gleam-playground/?s=A4VwRgBAZgdhC2BDAljAFASggbwFAQgBsBTAF2gHsLcBfIA%3D

I got to this when thinking non-gleam:

1. initialize variable
2. rebind


```gleam
// invalid pseudo code

let foo
foo = 5
```

File: compiler-core/src/parse/tests.rs [hunk 1]
```diff
@@ -232,3 +232,15 @@ fn anonymous_function_labeled_arguments() {
         }
     );
 }
+
+#[test]
+fn missing_let_binding() {
+    assert_error!(
+        "let bar = 42;
```
Comment (inoas) [inline compiler-core/src/parse/tests.rs:239]: AFAIK the semicolons are tolerated but removed by the formatter and shouldn't be there, but I could be wrong.

Comment (Kn99HN) [reply compiler-core/src/parse/tests.rs:239]: Gotcha.

**Comment by Kn99HN** [comment]
> "="

Thanks for the comment. I also just realized that the case when `foo: Int = 32` wouldn't return an error. I'll look more into it

**Comment by Kn99HN** [comment]
> I also noticed that this error message isn't optimal:
> 
> ```gleam
> pub fn main() {
>   let foo
> }
> ```
> 
> ```
> error: Syntax error
>   ┌─ ./src/main.gleam:3:1
>   │
> 3 │ }
>   │ ^ I was not expecting this.
> 
> Expected one of: "="
> ```
> 
> https://johndoneth.github.io/gleam-playground/?s=A4VwRgBAZgdhC2BDAljAFASggbwFAQgBsBTAF2gHsLcBfIA%3D
> 
> I got to this when thinking non-gleam:
> 
>     1. initialize variable
> 
>     2. rebind
> 
> 
> ```gleam
> // invalid pseudo code
> 
> let foo
> foo = 5
> ```

So I think in this case, the error message is not wrong but it is not the most accurate location. So it can be confusing. My approach is to make it more explicit error by relying on pattern.location(). Please let me know your thoughts on this.

File: compiler-core/src/parse/error.rs [hunk 2]
```diff
@@ -211,6 +220,7 @@ pub enum ParseErrorType {
     NoConstructors,            // A type "A {}" must have at least one constructor
     NoCaseClause,              // a case with no clauses
     NoExpression, // between "{" and "}" in expression position, there must be an expression
+    NoLetBinding, // Bindings and rebinds always require let and must always bind to a value.
```
Comment (Kn99HN) [inline compiler-core/src/parse/error.rs:223]: Rename "InvalidAssignmentStatement" to "NoLetBinding" for more precise error.

**Comment by Kn99HN** [comment]
> 

Do you want the snapshot test be in the same file as `parse/test.rs`?

**Comment by lpil** [comment]
That would be good, although if there is not already a test macro for parse errors feel free to put it in one of the test modules that does have an error macro.

**Review by lpil — APPROVED** [review]
Thank you! Great work!

**Pull request merged**